### ISSUE TITLE
Add except: "after-single-line-comment" option to *-empty-line-after rules

### ISSUE THREAD

**Issue by ntwb** [stylelint/stylelint#1351]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

An an enhancement feature request for blank new lines after section comments

> Which rule, if any, is this issue related to?

Related to: [`comment-empty-line-before`](https://github.com/stylelint/stylelint/tree/master/src/rules/comment-empty-line-before)
> What CSS is needed to reproduce this issue?

The following patterns are _not_ considered warnings:
```css
/**
* #.# Section title
*
* Description of section, whether or not it has media queries, etc.
*/

a { color: pink;}

/* This is a comment about this selector */
a { color: pink;}
```

The following patterns are considered warnings:
```css
/**
* #.# Section title
*
* Description of section, whether or not it has media queries, etc.
*/
a { color: pink;}
/* This is a comment about this selector */
a { color: pink;}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {"comment-empty-line-before": "always",
  "rules": {"comment-empty-line-after": "always",
  }
}
```

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Not directly, but should take into account all PostCSS syntaxes

**Comment by jeddy3** [comment]
Use the following to enforce this code style:

- `at-rule-empty-line-before`
- `rule-nested-empty-line-before`
- `rule-non-nested-empty-line-before`

These rules (along with `comment-empty-line-before`) have been designed to push *themselves* away from their *preceding* nodes. Adding additional `*-empty-line-after` rules, which push *other nodes away*, will lead to conflict.

**Comment by ntwb** [comment]
I'm already using each of these 3 rules, and currently with each the option `ignore: ["after-comment"],`

As such that configuration doesn't quite satisfy exactly what I'm looking for:

What I'd like to be able to enforce is the following CSS:
```css
/**
* #.# Section title
*
* Description of section, whether or not it has media queries, etc.
*/

a { color: pink;}

/* This is a comment about this selector */
a { color: pink;}
```

- 1. That is a _new line_ after the multiline section title comment, i.e. this one:
```css
*/

a { color: pink;}
```

- 2. And at the same time _no new line_ after the single line comment, i.e. here:
```css
/* This is a comment about this selector */
a { color: pink;}
```

By using the option  `ignore: ["after-comment"]` with `rule-nested-empty-line-before` this satisfies number 2 example above, but _not_ the number 1 example, and inversely removing the `ignore: ["after-comment"]` option from `rule-nested-empty-line-before` will satisfy number 1 above but not number 2.

I'm now contemplating if a new option `after-multiline-comment` for each of those 3 rules would satisfy what I'm looking for, that way the `ignore` could be determined if it's from a multiline comment or a single line comment?

**Comment by ntwb** [comment]
This raises potentially another side issue, ensuring that there is not a blank line between the singlee line comment and the following selector it is related to:

The following patterns are not considered warnings:
```css
/* This is a comment about this selector */
a { color: pink;}
```


The following patterns are considered warnings:
```css
/* This is a comment about this selector */

a { color: pink;}
```

**Comment by davidtheclark** [comment]
I understand @jeddy3's concern that if we start introducing `*-empty-line-after` rules we open the door to a lot of rules that potentially contradict each other. I guess we could just do that and put it on the user to figure their config out.

If we do decide to go with it, how about `comment-empty-line-after` with the options: always, never, always-multi-line?

**Comment by jeddy3** [comment]
> I understand @jeddy3's concern that if we start introducing *-empty-line-after rules we open the door to a lot of rules that potentially contradict each other. I guess we could just do that and put it on the user to figure their config out.

A new option sgtm :) I don't think we need to consider new rules for this use-case...

@ntwb I think in your case it a `except: "after-single-line-comment"` option you're after. `except` because you want to *enforce* [the inverse of the primary option](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#determine-options). And `"after-single-line-comment`" as you want an empty line between things everything else like rules, at-rules and multi-line comments.

With the following config:

```js
"rule-non-nested-empty-line-before": ["always", { "except": ["after-single-line-comment"] }]
```

The following is valid:

```css
/**
* #.# Section title
*
* Description of section, whether or not it has media queries, etc.
*/

a { color: pink;}

/* This is a comment about this selector */
a { color: pink;}
```

Whereas the following will produce *two* errors:

```css
/**
* #.# Section title
*
* Description of section, whether or not it has media queries, etc.
*/
a { color: pink;}

/* This is a comment about this selector */

a { color: pink;}
```

This is why we originally developed the concept of `ignore` **and** `except`, rather than just `ignore` :)

**Comment by ntwb** [comment]
> I understand @jeddy3's concern that if we start introducing *-empty-line-after rules we open the door to a lot of rules that potentially contradict each other. I guess we could just do that and put it on the user to figure their config out.

Agreed, potential conflicting rules is not good, even with ~20,000+ tests, contradictions are never far away 😏 

> A new option sgtm :) I don't think we need to consider new rules for this use-case...

Agreed 😄 

> @ntwb I think in your case it is a `except: "after-single-line-comment"` option you're after. We're using `except`, rather than `ignore`, here because you want to enforce the inverse of the primary option. And `"after-single-line-comment"` as you want an empty line between everything else like rules, at-rules and multi-line comments.

Yes, this will work perfectly, thanks for helping flesh this out 👍

**Comment by ntwb** [comment]
Going to take a stab at adding this option myself 😱

**Comment by jeddy3** [comment]
> Going to take a stab at adding this option myself 

Excellent!

I think an example of everything you need [is here](https://github.com/stylelint/stylelint/blob/master/src/rules/rule-non-nested-empty-line-before/index.js#L56-L68) i.e. getting the previous node, checking if something is a single line string and working with the `except` options. Good luck :)

**Comment by jeddy3** [comment]
Done in #1543
